lib/ash/filter/filter.ex

```diff
@@ -857,14 +857,10 @@ defimpl Inspect, for: Ash.Filter do
   import Inspect.Algebra
   import Ash.Filter.InspectHelpers
 
-  def inspect(%Ash.Filter{impossible?: impossible, not: not_filter} = filter, opts)
-      when not is_nil(not_filter) do
-    impossible =
-      if impossible do
-        "X"
-      else
-        ""
-      end
+  defguardp is_empty(val) when is_nil(val) or length(val) == 0
```

Comment:
`val == [] or val == %{}` would be preferable, I think

Reply:
for performance.

Reply:
Does it actually matter? `length` should run in constant time right? Unless there's a minor perf difference in how it has to resolve the type first

Reply:
I don't see how `length()` would run in constant time. pretty sure its a traversal in Elixir.

Reply:
https://github.com/elixir-lang/elixir/issues/6374

Reply:
Seems possible they might special case `length(thing) == 0` to `thing == []` though.

Reply:
Ah yeah, I got `*_size` and `length` backwards. Yours is cleaner than what I had with `(is_map(...) and map_size(...) == 0) or length(...) == 0` though, so let's go with your proposal!